src/site/content/en/reliable/two-way-communication-guide/index.md

```diff
@@ -0,0 +1,400 @@
+---
+layout: post 
+title: "Two way communication" 
+authors:
+  - demianrenzulli
+  - andrewguan 
+date: 2020-12-04 
+description: | 
+    How establish a two-way communication channel between the page and the service worker. 
+tags:
+  - blog
+  - service-worker
+  - web-worker
+---
+
+In some cases, a web app might need to establish a **two-way** communication channel between the
+page and the service worker.
+
+For example: in a podcast PWA one could build a feature to let the user [download episodes for
+offline consumption](https://web.dev/app-like-pwas/#proactive-background-downloading) and allow the
+service worker to keep the page regularly informed about the progress, so the [main
+thread](https://developer.mozilla.org/en-US/docs/Glossary/Main_thread) can update the UI.
+
+In this guide we'll explore the different ways of implementing a **two-way** communication between
+the [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) and [service
+worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) context, by exploring
+different APIs, the [Workbox library](https://developers.google.com/web/tools/workbox), as well as
+some advanced cases.
+
+<figure class="w-figure">
+  <img src="two-way-communication-diagram.png"
+       width="800"
+       alt="Diagram showing a service worker and the page exchanging messages.">
+</figure>
+
+## Using Workbox
+
+[Workbox-window](https://developers.google.com/web/tools/workbox/modules/workbox-window) is a set of
+modules of the [Workbox library](https://developers.google.com/web/tools/workbox) that are intended
+to run in the window context. The [Workbox
+class](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-window.Workbox)
+provides a `messageSW()` method, to `postMessage()` the instance's registered service worker and
+await a response.
+
+The following page code, creates a new `Workbox` instance and sends a message to the service worker
+to obtain its version. The result of the operation is logged via console:
+
+```javascript
+const wb = new Workbox('/sw.js'); 
+wb.register(); 
+
+const swVersion = await wb.messageSW({type: 'GET_VERSION'}); 
+console.log('Service Worker version:', swVersion);
+```
+
+The service worker implements a message listener on the other end, and responds to the registered
+service worker:
+
+```javascript
+const SW_VERSION = '1.0.0';
+
+addEventListener('message', (event) => {
+  if (event.data.type === 'GET_VERSION') {
+    event.ports[0].postMessage(SW_VERSION);
+  }
+});
+```
+
+Under the hood the library uses a browser API that we'll review in the next section: [Message
+Channel](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API), but abstracts many
+implementation details, making it easier to use, while leveraging the [wide browser
+support](https://caniuse.com/?search=messagechannel) this API counts.
```

Comment:
```suggestion
support](https://caniuse.com/mdn-api_messagechannel_port1) this API has.
```
`?search` URLs on Can I Use are unstable, click the hashtag to get a stable link